zebra-network/src/protocol/external/codec.rs

```diff
@@ -674,6 +676,52 @@ impl Codec {
     fn read_filterclear<R: Read>(&self, mut _reader: R) -> Result<Message, Error> {
         Ok(Message::FilterClear)
     }
+
+    /// Given the reader, deserialize the transaction in the rayon thread pool.
+    #[allow(clippy::unwrap_in_result)]
+    fn deserialize_transaction_spawning<R: Read + std::marker::Send>(
+        reader: R,
+    ) -> Result<Transaction, Error> {
+        let mut result = None;
+
+        // Correctness: Do CPU-intensive work on a dedicated thread, to avoid blocking other futures.
+        //
+        // Since we use `block_in_place()`, other futures running on the connection task will be blocked:
+        // https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html
+        //
+        // We can't use `spawn_blocking()` because:
+        // - The `reader` has a lifetime (but we could replace it with a `Vec` of message data)
+        // - There is no way to check the blocking task's future for panics
+        tokio::task::block_in_place(|| {
+            rayon::in_place_scope_fifo(|s| {
+                s.spawn_fifo(|_s| result = Some(Transaction::zcash_deserialize(reader)))
+            })
+        });
+
+        result.expect("scope has already finished")
+    }
+
+    /// Given the reader, deserialize the block in the rayon thread pool.
+    #[allow(clippy::unwrap_in_result)]
+    fn deserialize_block_spawning<R: Read + std::marker::Send>(reader: R) -> Result<Block, Error> {
```

Comment:
To avoid duplication we could have a single `deserialize_spawning()` and pass the functions as arguments. Arguments will be `Block::zcash_deserialize(reader)` or `Transaction::zcash_deserialize(reader)` however we can do that later.